lib/auth/ec2_join.go

```diff
@@ -0,0 +1,861 @@
+	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/jonboulle/clockwork"
+	"go.mozilla.org/pkcs7"
+)
+
+type ec2Client interface {
+	DescribeInstances(ctx context.Context, params *ec2.DescribeInstancesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type ec2ClientKey struct{}
+
+func ec2ClientFromContext(ctx context.Context) (ec2Client, bool) {
+	ec2Client, ok := ctx.Value(ec2ClientKey{}).(ec2Client)
+	return ec2Client, ok
+}
+
+// ec2ClientFromConfig returns a new ec2 client from the given aws config, or
+// may load the client from the passed context if one has been set (for tests).
+func ec2ClientFromConfig(ctx context.Context, cfg aws.Config) ec2Client {
+	ec2Client, ok := ec2ClientFromContext(ctx)
+	if ok {
+		return ec2Client
+	}
+	return ec2.NewFromConfig(cfg)
+}
+
+// NodeIDFromIID returns the node ID that must be used for nodes joining with
+// the given Instance Identity Document.
+func NodeIDFromIID(iid *imds.InstanceIdentityDocument) string {
+	return iid.AccountID + "-" + iid.InstanceID
+}
+
+// checkEC2AllowRules checks that the iid matches at least one of the allow
+// rules of the given token.
+func checkEC2AllowRules(ctx context.Context, iid *imds.InstanceIdentityDocument, token types.ProvisionToken) error {
+	allowRules := token.GetAllowRules()
+	for _, rule := range allowRules {
+		if len(rule.AWSAccount) > 0 {
+			if rule.AWSAccount != iid.AccountID {
+				continue
+			}
+		}
+		if len(rule.AWSRegions) > 0 {
+			matchingRegion := false
+			for _, region := range rule.AWSRegions {
+				if region == iid.Region {
+					matchingRegion = true
+					break
+				}
+			}
+			if !matchingRegion {
+				continue
+			}
+		}
+		// iid matches this allow rule. Check if it is running.
+		return trace.Wrap(checkInstanceRunning(ctx, iid.InstanceID, iid.Region, rule.AWSRole))
+	}
+	return trace.AccessDenied("instance did not match any allow rules")
+}
+
+func checkInstanceRunning(ctx context.Context, instanceID, region, IAMRole string) error {
+	awsClientConfig, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	awsClientConfig.Region = region
+
+	// assume the configured IAM role if necessary
+	if IAMRole != "" {
+		stsClient := sts.NewFromConfig(awsClientConfig)
+		creds := stscreds.NewAssumeRoleProvider(stsClient, IAMRole)
+		awsClientConfig.Credentials = aws.NewCredentialsCache(creds)
+	}
+
+	ec2Client := ec2ClientFromConfig(ctx, awsClientConfig)
+
+	output, err := ec2Client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []string{instanceID},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(output.Reservations) == 0 || len(output.Reservations[0].Instances) == 0 {
+		return trace.AccessDenied("")
```

Comment:
Could we provide some error context here and below like "failed to get an instance state"

Reply:
yes, added